cli/standalone/serialization.rs

```diff
@@ -0,0 +1,642 @@
+      npm_snapshot.map(serialize_npm_snapshot).unwrap_or_default();
+    write_bytes_with_len(&mut bytes, &npm_snapshot);
+  }
+  // 3. Remote modules
+  {
+    let update_index = bytes.len();
+    bytes.extend_from_slice(&(0_u64).to_le_bytes());
+    let start_index = bytes.len();
+    remote_modules.write(&mut bytes)?;
+    let length = bytes.len() - start_index;
+    let length_bytes = (length as u64).to_le_bytes();
+    bytes[update_index..update_index + length_bytes.len()]
+      .copy_from_slice(&length_bytes);
+  }
+  // 4. VFS
+  {
+    let (vfs, vfs_files) = vfs.into_dir_and_files();
+    let vfs = serde_json::to_string(&vfs)?;
+    write_bytes_with_len(&mut bytes, vfs.as_bytes());
+    let vfs_bytes_len = vfs_files.iter().map(|f| f.len() as u64).sum::<u64>();
+    bytes.extend_from_slice(&vfs_bytes_len.to_le_bytes());
+    for file in &vfs_files {
+      bytes.extend_from_slice(file);
+    }
+  }
+
+  // write the magic bytes at the end so we can use it
+  // to make sure we've deserialized correctly
+  bytes.extend_from_slice(MAGIC_BYTES);
+
+  Ok(bytes)
+}
+
+pub struct DeserializedDataSection {
+  pub metadata: Metadata,
+  pub npm_snapshot: Option<ValidSerializedNpmResolutionSnapshot>,
+  pub remote_modules: RemoteModulesStore,
+  pub vfs_dir: VirtualDirectory,
+  pub vfs_files_data: &'static [u8],
+}
+
+pub fn deserialize_binary_data_section(
+  data: &'static [u8],
+) -> Result<Option<DeserializedDataSection>, AnyError> {
+  fn read_bytes_with_len(input: &[u8]) -> Result<(&[u8], &[u8]), AnyError> {
+    let (input, len) = read_u64(input)?;
+    let (input, data) = read_bytes(input, len as usize)?;
+    Ok((input, data))
+  }
+
+  fn read_magic_bytes(input: &[u8]) -> Result<(&[u8], bool), AnyError> {
+    if input.len() < MAGIC_BYTES.len() {
+      bail!("Unexpected end of data. Could not find magic bytes.");
+    }
+    let (magic_bytes, input) = input.split_at(MAGIC_BYTES.len());
+    if magic_bytes != MAGIC_BYTES {
+      return Ok((input, false));
+    }
+    Ok((input, true))
+  }
+
+  let (input, found) = read_magic_bytes(data)?;
+  if !found {
+    return Ok(None);
+  }
+
+  // 1. Metadata
+  let (input, data) = read_bytes_with_len(input).context("reading metadata")?;
+  let metadata: Metadata =
+    serde_json::from_slice(data).context("deserializing metadata")?;
+  // 2. Npm snapshot
+  let (input, data) =
+    read_bytes_with_len(input).context("reading npm snapshot")?;
+  let npm_snapshot = if data.is_empty() {
+    None
+  } else {
+    Some(deserialize_npm_snapshot(data).context("deserializing npm snapshot")?)
+  };
+  // 3. Remote modules
+  let (input, data) =
+    read_bytes_with_len(input).context("reading remote modules data")?;
+  let remote_modules =
+    RemoteModulesStore::build(data).context("deserializing remote modules")?;
+  // 4. VFS
+  let (input, data) = read_bytes_with_len(input).context("vfs")?;
+  let vfs_dir: VirtualDirectory =
+    serde_json::from_slice(data).context("deserializing vfs data")?;
```

Comment:
FYI in `deno_core` we use https://crates.io/crates/bincode to serialize snapshot, maybe it would be useful in this PR?

Reply:
Yeah, I think we should do it in the future.